plenum/server/consensus/checkpoint_service.py

```diff
@@ -1,9 +1,351 @@
+
+    def isPpSeqNoBetweenWaterMarks(self, ppSeqNo: int):
+        return self._data.low_watermark < ppSeqNo <= self._data.high_watermark
+
+    def update_watermark_from_3pc(self):
+        if (self.last_ordered_3pc is not None) and (self.last_ordered_3pc[0] == self.view_no):
+            self._logger.info("update_watermark_from_3pc to {}".format(self.last_ordered_3pc))
+            self._data.low_watermark = self.last_ordered_3pc[1]
+        else:
+            self._logger.info("try to update_watermark_from_3pc but last_ordered_3pc is None")
+
+    def _remove_stashed_checkpoints(self, till_3pc_key=None):
+        """
+        Remove stashed received checkpoints up to `till_3pc_key` if provided,
+        otherwise remove all stashed received checkpoints
+        """
+        if till_3pc_key is None:
+            self.stashed_recvd_checkpoints.clear()
+            self._logger.info('{} removing all stashed checkpoints'.format(self))
+            return
+
+        for view_no in list(self.stashed_recvd_checkpoints.keys()):
+
+            if view_no < till_3pc_key[0]:
+                self._logger.info('{} removing stashed checkpoints for view {}'.format(self, view_no))
+                del self.stashed_recvd_checkpoints[view_no]
+
+            elif view_no == till_3pc_key[0]:
+                for (s, e) in list(self.stashed_recvd_checkpoints[view_no].keys()):
+                    if e <= till_3pc_key[1]:
+                        self._logger.info('{} removing stashed checkpoints: '
+                                          'viewNo={}, seqNoStart={}, seqNoEnd={}'.
+                                          format(self, view_no, s, e))
+                        del self.stashed_recvd_checkpoints[view_no][(s, e)]
+                if len(self.stashed_recvd_checkpoints[view_no]) == 0:
+                    del self.stashed_recvd_checkpoints[view_no]
+
+    def reset_checkpoints(self):
+        # That function most probably redundant in PBFT approach,
+        # because according to paper, checkpoints cleared only when next stabilized.
+        # Avoid using it while implement other services.
+        self._checkpoint_state.clear()
+        self._data.checkpoints.clear()
+        self._data.stable_checkpoint = 0
+
+    def _set_stable_checkpoint(self, end_seq_no):
+        if not list(self._data.checkpoints.irange_key(end_seq_no, end_seq_no)):
+            raise LogicError('Stable checkpoint must be in checkpoints')
+        self._data.stable_checkpoint = end_seq_no
+
+        self._data.checkpoints = \
+            SortedListWithKey([c for c in self._data.checkpoints if c.seqNoEnd >= end_seq_no],
+                              key=lambda checkpoint: checkpoint.seqNoEnd)
+
+        for start, end in list(self._checkpoint_state.keys()):
+            if end < end_seq_no:
+                self._checkpoint_state.pop((start, end))
+        self._clear_batch_till_seq_no(end_seq_no)
+
+    def _clear_batch_till_seq_no(self, seq_no):
```

Comment:
I think that eventually this needs to be called from OrderingService (in `gc`) when receiving `Cleanup` message on internal bus and cleaning up till stable Checkpoint.